### DISCUSSION TITLE
hydra-zen Gotchya List

### DISCUSSION THREAD

**Discussion by jgbos** [mit-ll-responsible-ai/hydra-zen#142] [General]
A list of "gotchyas":

## Configuring "Interactive" Functions

Be careful configuring functions defined in interactive modes for reproducibility.

```python
>> from hydra_zen import builds, to_yaml
>> def f(p=2):
..    return p
>> config = builds(f)
>> print(to_yaml(config))
_target_: __main__.f
_recursive_: true
_convert_: none
```

## OmegaConf Interpolations

Be careful in setting OmegaConf interpolations as defaults.

```python
>> from hydra_zen import builds, instantiate
>> def f(p=2):
..    return p
>> config = builds(f, p="${experiment.p})
>> instantiate(config)
InterpolationKeyError: Interpolation key 'experiment.p' not found
```

## Config Defaults for Objects
If you need to override "objects" in a config you must set those objects to `None` in your configuration.

```python
>> from hydra_zen import builds, to_yaml
>> from hydra_zen.experimental import hydra_run
>> from hydra.core.config_store import ConfigStore
>> cs = ConfigStore.instance()
>> def f(p=2):
..    return p
>> cs.store(group="myfun", name="f", node=builds(f))
>> def g(q=3):
..    return q
>> cs.store(group="myfun", name="g", node=builds(g))
>> config = dict(myfun=builds(f))
>> hydra_run(config, instantiate, overrides=["+myfun=g"])
Merge error : Builds_g is not a subclass of Builds_f. value: {'_target_': '__main__.g', '_recursive_': True, '_convert_': 'none'}
    full_key: 
    object_type=Config
```

## Submitit-Launcher and Environment
If you try to submit jobs from an interactive node on Slurm you should first remove the environment variables for Slurm since submitit copies the current environment.  

```python
for k in os.environ.keys():
    if "SLURM" in k:
        del os.environ[k]
```

**Comment by addisonklinke** [comment]
Aside from `InterpolationKeyError` when instantiating an object / running a Hydra script, it feels like class init signatures should not use OmegaConf interpolations. Otherwise, they really cannot be used outside the context of a Hydra script (i.e. elsewhere in the library). 

What do you all think about an `interpolations` class attribute as a workaround? This would allow the class to be used in a normal Python way as well as with Hydra instantiation

```python
from hydra_zen import builds, instantiate, to_yaml
from omegaconf import OmegaConf


class Model:
    
    interpolations = {
        'crop_height': '${dataset:crop_height}',
        'crop_width': '${dataset:crop_width}'}
    
    def __init__(self, crop_height: int = 224, crop_width: int = 224):
        pass


# Basic Python (could be used elsewhere in broader library)
python_model = Model()

# Hydra script usage (with registered dummy interpolation)
ModelConf = builds(Model, **Model.interpolations)
OmegaConf.register_new_resolver('dataset', lambda _: 224)
print(to_yaml(ModelConf))
hydra_model = instantiate(ModelConf)
```

**Reply by rsokl** [reply]
(Sorry for the delayed response! Just got back from vacation 🏖️ )

>  it feels like class init signatures should not use OmegaConf interpolations

I definitely agree with this. hydra-zen is meant to encourage users to write Python code without needing to pay heed to Hydra/OmegaConf. It would be especially gross if you could only instantiate a class by using the `instantiate` function.

> What do you all think about an interpolations class attribute as a workaround? 

I like that this bundles the relevant interpolation fields with `Model` without them getting in the way of just using `Model`. If you are planning on developing code in a consistent "Hydra-aware" way, where you make it easy to configure your code for a Hydra-based application, then I think your approach is a good one. 


### A quick tangent about meta-fields:

You might be interested in leveraging [meta-fields](https://mit-ll-responsible-ai.github.io/hydra-zen/generated/hydra_zen.builds.html#meta-field) which is a feature uniquely enabled by hydra-zen. This would enable you to control the height & width from a single interpolated field that is exposed from the model's config, without needing to register a resolver. E.g.

```python
class Model:
    
    interpolations = {
        'crop_height': '${.size}',
        'crop_width': '${.size}',
        'zen_meta': dict(size=224)}
    
    def __init__(self, crop_height: int = 224, crop_width: int = 224):
        self.height = crop_height
        self.width = crop_width

    def __repr__(self): return str((self.height, self.width))
```

```python
>>> ModelConf = builds(Model, **Model.interpolations)
>>> instantiate(ModelConf, size=12)
(12, 12)
```

However, a downside to this approach is that `.size` is a relative interpolation, not a global one. I.e. if your `ModelConf` is nested within another config, you'd have to configure `size` accordingly. E.g. for the config

```python
from hydra_zen import make_config

Conf = make_config(level1=ModelConf)
```
 
you would only be able to overwrite `size` via: `level1.size=12` from the CLI. Your method of using a globally-available resolver would make it easier to control the crop size, especially if you use the same resolver across multiple configs. Another downside to using meta-fields is that the resulting dataclasses can only be instantiated if hydra-zen is installed. 

Just wanted to make sure that this feature is on your radar!

**Reply by addisonklinke** [reply]
Good to know you think the `interpolations` attribute is a nice approach for Hydra-based applications

Is the behavior of `cfg = builds(obj, **obj.get('interpolations', {}))` something you'd consider adding to `builds`? In case of name collisions where `interpolations` is a pre-existing attribute that's unrelated to Hydra, I think some type-checking might suffice. For instance, `interpolations` must be a dict with string values containing the OmegaConf `${...}` syntax. Alternatively, the official name could be `hydra_zen_interpolations` which would make name collisions very unlikely. Either way, I think it would be cool to use the library to encourage a best-practice here

Thanks for pointing out the meta-fields approach! That's useful to know, but I think in my specific application there will be more nesting and shared resolvers so I'd rather stick to non-relative interpolations

**Reply by rsokl** [reply]
> Is the behavior of `cfg = builds(obj, **obj.get('interpolations', {}))` something you'd consider adding to `builds`?

My main concern with this approach is that is encourages a design such that the end-user's config creation code is somewhat cryptic. I.e. I prefer the form

```python
cfg = builds(obj, **obj.interpolations)
```

because it makes clear to the reader that the config is being populated by a dictionary stored on `obj` itself. Whereas

```python
cfg = builds(obj)
```

requires the user to know both that `builds(<target>)` has special behavior when `.interpolations` is present *and* that `obj` has such an attribute. Given that using Hydra can already be somewhat confusing for newcomers, I think that simple, readable config-creation code is quite important.

Of course, you are free to ship/use your own extension of `builds`, if you would like to regularly leverage this sort of feature.

```python
def auto_builds(obj, *args, **kwargs):
    builtin_kwargs = obj.get('interpolations', {})
    if isinstance(builtin_kwargs, dict):
        kwargs.update(builtin_kwargs)
    return builds(obj, *args, **kwargs)
```

That all being said, I appreciate you asking this question because it makes me think about how hydra-zen might design/promote this sort of functionality in a standardized for library authors to leverage. I have some thoughts on this that are stewing, and I will share them with you asap 😄

**Reply by addisonklinke** [reply]
Yeah I understand the concerns around the cryptic behavior. Keep me posted what you come up with - like you mentioned I think it's a good opportunity to define a community best-practice for this so everyone doesn't go around rolling their own extensions of `builds` with slightly different approaches

**Reply by addisonklinke** [reply]
Thinking about it some more, I might lean towards an interface like this

```python
class MyClass:

    __hydra__ = {
        'interpolations': {},
        'defaults': []}
```

I like that it indicates an important internal, object-orientated property similar to `__len__` and others. In that regard it's also unlikely to run into namespace collisions

**Reply by rsokl** [reply]
Right, a nice thing about a name like `__hydra__` is that it also makes it pretty clear to readers that this is simply extending the functionality of the class – in the same way that `__len__` enables a class to work with `len`, `__hydra__` could enable compatibility with `builds`.

I want to make sure I understand your ultimate goals here. Is it correct to say that you want to:
- Design a standardized set of config fields that can be leveraged across your Hydra configs via interpolation
- Incorporate those interpolated fields, and other default config info, into the designs of the classes themselves.
- Be able to use `builds` to trivially convert any given object in your library to a structured config with the appropriate interpolated fields & defaults already set
- Achieve all of this in a standardized way, as facilitated by hydra-zen, that other libraries can leverage as well

Is that correct? Anything missing?

**Reply by addisonklinke** [reply]
Yes that's a great summary. My main takeaway - similar to how the type annotations/init signature of the class drive the behavior of `builds()` in a standardized way, we should have a similar interface to define the interpolations/defaults since those are frequently useful in Hydra configs, especially as the application gets more complex

The only thing missing is the concept of a `builds_decorator`, but that is already discussed on https://github.com/mit-ll-responsible-ai/hydra-zen/issues/93 so it seems best to leave them separate. Sorry I fell a bit behind on the discussion there, will get back to it after this!

**Comment by addisonklinke** [comment]
Similar to the interpolations thread above, I'm trying to figure out the best way to allow `builds()` to specify the defaults list for a specific config node. This is important so that selecting a node from a config group can adjust the defaults for other nodes/groups as mentioned in the core Hydra [docs](https://hydra.cc/docs/next/upgrades/1.0_to_1.1/defaults_list_override/#internaldocs-banner). Unfortunately, naively adding `builds(obj, defaults=[...])` raises an error

```
TypeError: Building: MyClass ..
The following unexpected keyword argument(s) was specified for MyClass via `builds`: defaults
```

Currently, the only option I can think of is

```python
from dataclasses import field
from hydra_zen import builds
from lib import MyClass

Config = builds(MyClass)
defaults_list = field(default_factory=lambda: getattr(MyClass, 'defaults', []))
Config.__dataclass_fields__.update({'defaults': defaults_list})
```

Although this has the same issue with cryptic behavior raised by @rsokl above

**Reply by rsokl** [reply]
@addisonklinke given that `defaults` is not meant to be passed to the target-object during instantiation, it makes sense to specify it as a "meta-field" to `builds`:

```python
builds(obj, zen_meta=dict(defaults=[...]))
```

This will eliminate the `TypeError` that you reported above, and will work with Hydra as-expected:

```python
from hydra_zen import builds, instantiate, launch

import hydra
from hydra.core.config_store import ConfigStore


class MyClass:
    def __init__(self, **kwargs):
        self.kwargs = kwargs


Default = builds(int)

cs = ConfigStore.instance()
cs.store(group="x", name="a", node=Default)


Conf = builds(MyClass, x=None, zen_meta=dict(defaults=["_self_", {"x": "a"}]))

cs.store(name="my_app", node=Conf)  

@hydra.main(config_path=None, config_name="my_app")
def task_function(cfg):
    print(instantiate(cfg).kwargs)
```

```python
>>> launch(Conf, task_function);
{'x': 0}
```

**Reply by rsokl** [reply]
@addisonklinke I edited my post to just convey the meta-field solution here. I moved the notion of adding `hydra_defaults` to [this issue](https://github.com/mit-ll-responsible-ai/hydra-zen/issues/253#issue-1192287725). I listed some preliminary pros and cons there.

Do you frequently specify `defaults` on targeted configs?

**Reply by addisonklinke** [reply]
Sounds good for the issue 👍🏻 

I do find specifying `defaults` on a targeted config useful when it is a config group and selecting it should influence the default behavior. For a more concrete (and ML-specific) example, you could checkout this discussion [thread](https://github.com/facebookresearch/hydra/discussions/2126) I added to Hydra's Github earlier today. I'm running into an error which I believe is related to Hydra package syntax, but even so, the use-case makes a lot of sense for the kind of applications I work on

**Reply by rsokl** [reply]
Gotchya, and thanks for the link. Your use case does make sense to me.

Right now I am leaning towards supporting `hydra_defaults`, but in the meantime let me know if you have any issues with the meta-field solution.

**Reply by addisonklinke** [reply]
I tested out the `zen_meta` option and was able to reproduced the error I got there, so looking good so far. Anything standing out to you about how to adjust the package directives so that override applies to the correct config group?

**Reply by rsokl** [reply]
> Anything standing out to you about how to adjust the package directives so that override applies to the correct config group?

Unfortunately, I tend to limit my config group usages to rather simple cases ([like this](https://mit-ll-responsible-ai.github.io/hydra-zen/tutorials/config_groups.html#putting-it-all-together)), so I don't have much insights when it comes to matters like `@_global_`.  

One thing to note, there seems to be a typo in your example. You store the group named `'crop'`:

```
cs.store(group='datamodule/prepper', name='crop', node=PrepConf)
```

but then your `defaults` refers to `'crops'`:


```
    defaults: List[Any] = field(default_factory=lambda: [
        {'override datamodule/prepper@_global_': 'crops'},  # FIXME what's the correct package syntax here?
    ])
```

**Reply by addisonklinke** [reply]
For reference, one of the Hydra devs replied on their thread with the correct package syntax

```
defaults: List[Any] = field(default_factory=lambda: [
    {'/datamodule/prepper@_group_': 'crop'},
])
```

**Reply by rsokl** [reply]
@addisonklinke FYI: https://github.com/mit-ll-responsible-ai/hydra-zen/pull/264

**Comment by addisonklinke** [comment]
@rsokl I'm running into the "Config Defaults for Objects" issue you mentioned. I wanted to have a default model optimizer while being able to select alternates via CLI overrides. I tried making the problematic config group have a default of `None` as mentioned, but that doesn't resolve the issue

```python
# Contents of demo.py

from dataclasses import dataclass, field
from functools import partial
from typing import Any, List

import hydra
from hydra.core.config_store import ConfigStore
from hydra_zen import builds
from omegaconf import OmegaConf, MISSING
from torch import optim


@dataclass
class Conf:
    defaults: List[Any] = field(default_factory=lambda: [
        {'model/optimizer': None}
    ])
    model: Any = MISSING


class BaseModel:

    def __init__(self, optimizer=partial(optim.Adam)):
        pass


def register_configs():
    AdadeltaConf = builds(optim.Adadelta, populate_full_signature=True, zen_partial=True)
    cs = ConfigStore.instance()
    cs.store('config', Conf)
    cs.store(group='model', name='base', node=builds(BaseModel, populate_full_signature=True))
    cs.store(group='model/optimizer', name='Adadelta', node=AdadeltaConf)


@hydra.main(config_name='config', config_path=None)
def main(cfg):
    resolved_cfg = OmegaConf.to_container(cfg, resolve=True, enum_to_str=True)
    print(OmegaConf.to_yaml(resolved_cfg))


if __name__ == '__main__':

    register_configs()
    main()
```

Works fine without overriding the optimizer (minus the fact that the recursive `builds()` doesn't populate the full signature of `Adam`)

```bash
$ python demo.py +model=base
model:
  _target_: __main__.BaseModel
  optimizer:
    _target_: torch.optim.adam.Adam
    _partial_: true
```

But you can't pick an alternative optimizer

```bash
$ python demo.py +model=base model/optimizer=Adadelta
In 'model/base': Validation error while composing config:
Merge error: PartialBuilds_Adam is not a subclass of PartialBuilds_Adadelta. value: {'_target_': 'torch.optim.adam.Adam', '_partial_': True}
    full_key: 
    object_type=Conf
```

**Reply by rsokl** [reply]
I am not sure why specifying the default of `None` isn't working here..  It looks like the combo of:

```python
class BaseModel:

    def __init__(self, optimizer=partial(optim.Adam)):
        pass

cs.store(group='model', name='base', node=builds(BaseModel, populate_full_signature=True))
```

leads to the creation of `optimizer=builds(Adam, zen_partial=True)` due to `def __init__(self, optimizer=partial(optim.Adam))`, which is a slick way to flesh out "auto-config" support. That being said, Hydra seems to be specifying the resulting optimizer config as the default/base config, and so overriding the config group requires the new config to be a super-type of the the base config.. I really wish Hydra could let you just disable this check. It would alleviate many headaches.

@jgbos  do you have any ideas for workarounds?

**Reply by jgbos** [reply]
I think the only solution is to remove the default `{'model/optimizer': None}`.  Is there a reason you need to specify it as `None`?  I think you will need to create a new config:
```python
def nonefn():
    return None
cs.store(name="none", group="model/optimizer", node=builds(nonefn))
```
and set `{'model/optimizer': "none"}`

**Reply by addisonklinke** [reply]
@jgbos I was trying `None` just based on the recommendation in the initial post here. When I switch to `builds(nonefn)` instead, there is still an error because `builds(BaseModel)` is trying to pull in the default Adam optimizer

```
In 'model/base': Validation error while composing config:
Merge error: PartialBuilds_Adam is not a subclass of Builds_nonefn. value: {'_target_': 'torch.optim.adam.Adam', '_partial_': True}
    full_key: 
    object_type=Conf
```

> I really wish Hydra could let you just disable this check

Strongly agree - there seem to be some pretty similar discussions around this use case
* https://github.com/facebookresearch/hydra/issues/2227
* https://github.com/facebookresearch/hydra/issues/2039

One [suggestion](https://github.com/facebookresearch/hydra/issues/2227#issuecomment-1227196949) in those threads is to leave the node empty to begin with (which is what I think @rsokl and I were going for with the `None` default). However this seems to be fundamentally incompatible with auto-populating a default for the `BaseModel` class via `builds()`. We definitely don't want to code `optimizer=partial(optim.Adam)` into the class definition only to have to repeat it in the CLI overrides

**Reply by jgbos** [reply]
@addisonklinke  ah, sorry.  I didn't catch the default in `BaseModel`.  This definitely falls in the gotchas with Hydra.  The best solutions is to not provide a default to `BaseModel` and set the default optimizer in the config instead (which can be `None`).  I just tried a bunch of options without success.
